### DISCUSSION TITLE
Using testing.User should not require Selenium

### DISCUSSION THREAD

**Discussion by marcuslimdw** [zauberzeug/nicegui#3510] [Ideas / Feature Requests]
### Description

`testing.__init__.py` contains this:

```py
from .screen import Screen
from .user import User
from .user_interaction import UserInteraction

__all__ = [
    'Screen',
    'User',
    'UserInteraction',
]
```

In `screen.py`, objects from Selenium are imported. This means that even if I only want to use `testing.User` for Python-native testing, I need to include Selenium as a dependency, which makes testing a lot more inconvenient.

I suggest that:

1. Imports that are used only for type checking be put in a `if TYPE_CHECKING` block.
2. `ImportErrors` encountered when performing functional imports be swallowed and some sort of flag (e.g. `_selenium_available`) be set to `False`. If this flag is `False` when a `Screen` is instantiated, an appropriate exception will be raised (e.g. `"You need to install Selenium to use Screen for testing"`).

(I can maybe commit to raising a PR for this, if my suggestion is the direction the maintainers want to go in?)

**Comment by rodja** [comment]
Thanks for the suggestion @marcuslimdw. Maybe the problem could also be solved by providing separate plugins for screen and user as I just tried to implement in #3511.

**Reply by marcuslimdw** [reply]
that would be possible, but as I alluded to in https://github.com/zauberzeug/nicegui/discussions/3508, would you be willing to consider the ability to instantiate a user entirely outside of a `pytest` plugin?

**Reply by rodja** [reply]
> ... instantiate a user entirely outside of a pytest plugin...

Let's discuss that over on #3508. I think the PR #3511 is a good enhancement on it's own because you do not need selenium at all anymore.

**Comment by falkoschindler** [comment]
With PR #3511 being released in version 1.4.35, I'll close this feature request as completed.

**Reply by marcuslimdw** [reply]
great, thank you! 🙏

**Reply by marcuslimdw** [reply]
@rodja, could we please reopen this? This does not work as expected because of `.venv/lib/python3.10/site-packages/nicegui/testing/__init__.py`:

```py
from .screen import Screen
from .user import User
from .user_interaction import UserInteraction

__all__ = [
    'Screen',
    'User',
    'UserInteraction',
]
```

Imports come about via this module, which means that even if I use the user plugin, the Selenium import is still triggered.

**Comment by falkoschindler** [comment]
How can we allow importing `User` without importing `Screen`?
Claude.ai recommends something like this:
```py
from __future__ import annotations
from typing import Any

def __getattr__(name: str) -> Any:
    if name == 'Screen':
        from .screen import Screen
        return Screen
    elif name == 'User':
        from .user import User
        return User
    elif name == 'UserInteraction':
        from .user_interaction import UserInteraction
        return UserInteraction
    else:
        raise AttributeError(f'module {__name__!r} has no attribute {name!r}')

__all__ = [
    'Screen',
    'User',
    'UserInteraction',
]
```
What do you think?

**Reply by falkoschindler** [reply]
Or we check for selenium first:

```py
import importlib.util

from .user import User
from .user_interaction import UserInteraction

__all__ = [
    'User',
    'UserInteraction',
]

if importlib.util.find_spec('selenium'):
    from .screen import Screen
    __all__.append('Screen')
```

**Reply by marcuslimdw** [reply]
My (rather roundabout) thoughts:

Custom `__getattr__`, IMO, is more of a last resort, because it's not as discoverable/understandable and causes problems for type checkers.

I do like the second option (checking for Selenium), save that using `importlib.util.find_spec` is a little brittle, because it introduces a dependency on Selenium specifically - what if (admittedly very unlikely) the project decides to use a different automation framework, or the user happens to have a `selenium.py` in their project? I suggest wrapping `from .screen import Screen` in a `try-except ImportError` block instead.

That said, I'm not sure how modifying `__all__` at runtime will interact with type checkers/linters, since those perform static analysis.

The only alternative I can think of is to remove everything from `__init__.py`, and instead require the user to import from submodules, which is inconvenient *and* backwards incompatible.

All in all, I suggest the "check for Selenium" option, except with a `try-except` instead. 

Nevertheless, thank you for your quick reply! 🙏

**Reply by falkoschindler** [reply]
Very interesting thoughts! I totally agree with your concerns about altering `__all__` at runtime. We shouldn't do that.
Using try-except over find_spec sounds reasonable. But wouldn't a selenium.py in the project directory cause the very same problem?

**Reply by marcuslimdw** [reply]
AIUI, no.

With `find_spec`:

1. `find_spec` identifies first-party `selenium.py` file
2. `if` takes the `True` branch (**false positive**!)
3. `from .screen import Screen` is executed
4. Import fails because the necessary members (e.g. `webdriver`, `ActionChains`, `By`) do not actually exist in `selenium.py`, raising a `ModuleNotFoundError`
5. This error is propagated to the user

With `try-except`:

1. `from .screen import Screen` is executed
2. Import fails because the necessary members (e.g. `webdriver`, `ActionChains`, `By`) do not actually exist in `selenium.py`, raising a `ModuleNotFoundError`
3. `except` branch in `__init__.py` is triggered and swallows the error (perhaps add a log statement here?)
4. User does what they want

(please note that I got the wrong error - it shouldn't be `ImportError`, but instead `ModuleNotFoundError`)

**Reply by rodja** [reply]
Ok. I just created pull request #3585. Does that work for you @marcuslimdw?

**Reply by marcuslimdw** [reply]
@rodja I'm good with that, but purely from a DX perspective, it might be helpful to have some sort of user feedback (if not logging, at least a `print` statement) for those who might not know that their environment doesn't have Selenium set up, since they'll get a somewhat cryptic error

thank you for your work on this!

**Reply by rodja** [reply]
Good point. I just created a dummy `Screen` class which raises an ImportError if accessed.

**Reply by rodja** [reply]
After some experiments I think the feedback will not show up in real use cases. The `Screen` class is only usable if you have the fixture. And to get the fixture you need `screen_plugin.py` or `plugin.py`. Both imports will directly import Screen raise the missing Selenium exception on startup. Or have I missed a scenario where you would instantiate a `Screen` directly?

**Reply by marcuslimdw** [reply]
That makes sense - I think you're right.

In that case, do you feel that the "can't import Selenium" error is sufficiently descriptive that no further work needs to be done here? Or do you think that in e.g. `screen_plugin.py`, the `import selenium` should itself be wrapped with a `try-except` to provide a slightly more obvious error to the user (i.e. Selenium is optional, but if you want to use this to test your code, you'll need it)?

IMO, it's not necessary because this is already documented in the NiceGUI docs, but 🤷

**Reply by rodja** [reply]
>  Or do you think that in e.g. screen_plugin.py, the import selenium should itself be wrapped with a try-except to provide a slightly more obvious error to the user

If someone uses `plugin.py` or `screen_plugin.py` I think the import error should show. Hence I rolled back the changes.
